Libraries/LibWeb/DOM/Document.cpp

```diff
@@ -4626,7 +4626,10 @@ void Document::update_for_history_step_application(GC::Ref<HTML::SessionHistoryE
             //    given document's relevant global object to fire an event named hashchange at document's relevant global object,
             //    using HashChangeEvent, with the oldURL attribute initialized to the serialization of oldURL and the newURL attribute
             //    initialized to the serialization of entry's URL.
-            if (old_url.fragment() != entry->url().fragment()) {
+            // NOTE: Treat null URL fragments as empty strings to ensure consistent handling when comparing them.
+            auto old_url_fragment = old_url.fragment().has_value() ? old_url.fragment() : String {};
```

Comment:
From digging into the history of this WPT test, I don't think this is the correct place to fix this, or at least - this is not the only place we should be changing. Based on https://github.com/whatwg/html/pull/7473, it seems like it was intended that this check happens in `Location::set_hash` and avoid the navigation altogether. Changing Location::set_hash fixes the WPT test which is linked.

I would need to dig into this some more, but this seems like either a spec bug, or something which we are implementing wrong in some other place.

Reply:
I have dug into this a bit further, I think this is a bug in the HTML spec in the location.hash setter. I think it should be specified to consider null as equal to empty in the setter when comparing equality. I'm not sure if this call site also needs to be adjusted.

Reply:
Thanks for the code review! You're absolutely right, it's not the appropriate place to fix this issue, the specs are not very clear on this point. I took a look at Chromium's code, and it turns out that the comparison is also performed while ignoring null values using `EqualIgnoringNullity` function in `Location::setHash()`: https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/frame/location.cc

I've moved the fix to Location, which also has the side effect of fixing the test for the popstate event.